File: ios-native/Leyne/Redesign/RedesignCommon.swift
// Shared primitives for the redesign screens (iOS): an SF Symbol wrapper,
// occupancy → symbol/label/colour resolution, a crowd dot, a circular icon
// button, and a couple of small reused modifiers.

import SwiftUI

/// SF Symbol rendered at a fixed point size / weight / colour. Pass the
/// ".fill" variant name for filled glyphs (mirrors Material Symbols FILL=1).
struct RDSym: View {
    let name: String
    var size: CGFloat
    var color: Color
    var weight: Font.Weight = .regular

    init(_ name: String, size: CGFloat, color: Color, weight: Font.Weight = .regular) {
        self.name = name
        self.size = size
        self.color = color
        self.weight = weight
    }

    var body: some View {
        Image(systemName: name)
            .font(.system(size: size, weight: weight))
            .foregroundStyle(color)
    }
}

/// Occupancy resolved for arrival rows: (symbol, label, colour).
func rdOcc(_ load: RDLoad, _ t: RDTokens) -> (symbol: String, label: String, color: Color) {
    switch load {
    case .seats: return ("figure.seated.side", "Seats available", t.bus)
    case .standing: return ("figure.stand", "Standing room", t.amber)
    case .packed: return ("person.3.fill", "Packed", t.mrt)
    }
}

func rdLoadColor(_ load: RDLoad, _ t: RDTokens) -> Color {
    switch load {
    case .seats: return t.bus
    case .standing: return t.amber
    case .packed: return t.mrt
    }
}

/// Small circular crowd / live dot.
struct RDDot: View {
    let color: Color
    var size: CGFloat = 7
    var body: some View {
        Circle().fill(color).frame(width: size, height: size)
    }
}

/// Circular outline icon button used in detail-screen headers.
struct RDCircleButton: View {
    let symbol: String
    var label: String? = nil          // VoiceOver label for this icon-only control
    var bordered: Bool = true
    var iconColor: Color? = nil
    var bg: Color? = nil
    var size: CGFloat = 42
    var iconSize: CGFloat = 21
    let t: RDTokens
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            ZStack {
                Circle().fill(bg ?? Color.clear)
                if bordered {
                    Circle().strokeBorder(t.outlineVariant, lineWidth: 1)
                }
                RDSym(symbol, size: iconSize, color: iconColor ?? t.onSurface)
            }
            .frame(width: size, height: size)
            .contentShape(Circle())
        }
        .buttonStyle(.plain)
        .accessibilityLabel(label ?? "")
    }
}

/// Rounded card surface used by the list screens.
struct RDCard<Content: View>: View {
    let t: RDTokens
    var radius: CGFloat = 22
    @ViewBuilder var content: () -> Content

    var body: some View {
        content()
            .background(t.scLow)
            .clipShape(RoundedRectangle(cornerRadius: radius, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: radius, style: .continuous)
                    .strokeBorder(t.outlineVariant, lineWidth: 1)
            )
    }
}

/// "then 11" → "Next 11 min"; nil when there's no following bus.
func rdNextLabel(_ then: String?) -> String? {
    guard let then else { return nil }
    let digits = then.filter(\.isNumber)
    return digits.isEmpty ? nil : "Next \(digits) min"
}

/// One live-arrival row, shared by the Home and Stop screens so they stay
/// identical: a flat grouped-list row (no card) — neutral bus badge, destination,
/// occupancy dot, and an ETA-dominant right side (green "Now" when arriving) with
/// the following bus beneath.
struct RDArrivalRow: View {
    let a: RDArrival
    let t: RDTokens
    let onTap: () -> Void

    var body: some View {
        let occ = rdOcc(a.load, t)
        let arriving = (Int(a.min) ?? 99) <= 0
        return Button(action: onTap) {
            HStack(spacing: 14) {
                Text(a.route)
                    .font(rdFont(17, .heavy)).foregroundStyle(t.onSurface)
                    .frame(minWidth: 46).frame(height: 38).padding(.horizontal, 8)
                    .background(t.scHigh).clipShape(RoundedRectangle(cornerRadius: 11, style: .continuous))
                VStack(alignment: .leading, spacing: 4) {
                    Text(a.dest).font(rdFont(16.5, .semibold)).foregroundStyle(t.onSurface).lineLimit(1)
                    HStack(spacing: 6) {
                        RDDot(color: occ.color, size: 7)
                        Text(occ.label).font(rdFont(12.5, .medium)).foregroundStyle(t.onVariant).lineLimit(1)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                VStack(alignment: .trailing, spacing: 2) {
                    if arriving {
                        Text("Now").font(rdFont(17, .heavy)).foregroundStyle(t.bus)
                    } else {
                        HStack(alignment: .firstTextBaseline, spacing: 2) {
                            Text(a.min).font(rdFont(24, .black)).foregroundStyle(t.onSurface)
                                .contentTransition(.numericText())
                            Text("min").font(rdFont(11, .bold)).foregroundStyle(t.onVariant)
                        }
                    }
                    if let next = rdNextLabel(a.then) {
                        Text(next).font(rdFont(10.5, .medium)).foregroundStyle(t.onVariant)
                    }
                }
            }
            .padding(.horizontal, 18).padding(.vertical, 16)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .animation(.easeInOut(duration: 0.35), value: a.min)
    }
}

extension View {
    /// Fills the available width and aligns content to the leading edge.
    func rdLeading() -> some View {
        frame(maxWidth: .infinity, alignment: .leading)
    }
}

File: lib/screens/redesign/redesign_common.dart
// Shared primitives for the redesign screens: a Material Symbols icon wrapper
// (variable fill/weight), occupancy → icon/label/colour resolution, and a few
// small reused chip/circle-button builders.

import 'package:flutter/material.dart';
import 'package:material_symbols_icons/symbols.dart';

import 'redesign_data.dart';
import 'redesign_theme.dart';

/// Wraps a tappable child in a Material ink surface so it shows the M3 ripple /
/// state layer. Use on the flat grouped-list rows (which sit on a transparent
/// surface) — the redesign was built with bare GestureDetectors and had no
/// ripple feedback anywhere, the most un-Material thing about it.
Widget rdInk({
  required VoidCallback? onTap,
  required Widget child,
  BorderRadius? borderRadius,
}) {
  return Material(
    color: Colors.transparent,
    child: InkWell(onTap: onTap, borderRadius: borderRadius, child: child),
  );
}

/// Material Symbols Rounded icon with the variable-font axes the design uses.
class RdIcon extends StatelessWidget {
  const RdIcon(
    this.icon, {
    super.key,
    required this.size,
    required this.color,
    this.fill = 0,
    this.weight = 400,
  });

  final IconData icon;
  final double size;
  final Color color;
  final double fill; // 0 = outline, 1 = filled
  final double weight;

  @override
  Widget build(BuildContext context) {
    return Icon(
      icon,
      size: size,
      color: color,
      fill: fill,
      weight: weight,
      opticalSize: size,
    );
  }
}

/// Occupancy resolved for the arrival rows: (icon, label, colour).
({IconData icon, String label, Color color}) rdOcc(RdLoad load, RdTokens t) {
  switch (load) {
    case RdLoad.seats:
      return (icon: Symbols.airline_seat_recline_normal, label: 'Seats available', color: t.bus);
    case RdLoad.standing:
      return (icon: Symbols.accessibility_new, label: 'Standing room', color: t.amber);
    case RdLoad.packed:
      return (icon: Symbols.groups, label: 'Packed', color: t.mrt);
  }
}

/// Crowd dot colour for the bus / amber / mrt roles.
Color rdLoadColor(RdLoad load, RdTokens t) => switch (load) {
      RdLoad.seats => t.bus,
      RdLoad.standing => t.amber,
      RdLoad.packed => t.mrt,
    };

/// 8px circular dot used throughout (crowd / live indicators).
class RdDot extends StatelessWidget {
  const RdDot(this.color, {super.key, this.size = 7});
  final Color color;
  final double size;

  @override
  Widget build(BuildContext context) {
    return Container(
      width: size,
      height: size,
      decoration: BoxDecoration(color: color, shape: BoxShape.circle),
    );
  }
}

/// Circular outline icon button used in detail-screen headers.
class RdCircleButton extends StatelessWidget {
  const RdCircleButton({
    super.key,
    required this.icon,
    required this.onTap,
    this.label,
    this.bordered = true,
    this.iconColor,
    this.bg,
    this.fill = 0,
    this.size = 42,
    this.iconSize = 21,
  });

  final IconData icon;
  final VoidCallback onTap;
  final String? label; // TalkBack label for this icon-only control
  final bool bordered;
  final Color? iconColor;
  final Color? bg;
  final double fill;
  final double size;
  final double iconSize;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    return Semantics(
      button: true,
      label: label,
      child: Material(
        color: bg ?? Colors.transparent,
        shape: bordered
            ? CircleBorder(side: BorderSide(color: t.outlineVariant))
            : const CircleBorder(),
        clipBehavior: Clip.antiAlias,
        child: InkResponse(
          onTap: onTap,
          radius: size / 2 + 4,
          child: SizedBox(
            width: size,
            height: size,
            child: Center(
              child: RdIcon(icon, size: iconSize, color: iconColor ?? t.onSurface, fill: fill),
            ),
          ),
        ),
      ),
    );
  }
}

/// Standard detail-screen top bar padding (status bar handled by SafeArea).
const EdgeInsets kRdHeaderPad = EdgeInsets.fromLTRB(10, 6, 10, 0);
